op-node/sources/fallback_client.go

```diff
@@ -0,0 +1,243 @@
+package sources
+
+import (
+	"context"
+	"fmt"
+	"github.com/ethereum-optimism/optimism/op-node/client"
+	"github.com/ethereum-optimism/optimism/op-node/eth"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+	"math/big"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// FallbackClient is an RPC client, it can automatically switch to the next l1 endpoint
+// when there is a problem with the current l1 endpoint
+// and automatically switch back after the first l1 endpoint recovers.
+type FallbackClient struct {
+	// firstRpc is created by the first of the l1 urls, it should be used first in a healthy state
+	firstRpc          client.RPC
+	urlList           []string
+	rpcInitFunc       func(url string) (client.RPC, error)
+	lastMinuteFail    atomic.Int64
+	currentRpc        client.RPC
+	currentIndex      int
+	mx                sync.Mutex
+	log               log.Logger
+	isInFallbackState bool
+	subscribeFunc     func() (event.Subscription, error)
+	l1HeadsSub        *ethereum.Subscription
+	l1ChainId         *big.Int
+	l1Block           eth.BlockID
+	ctx               context.Context
+}
+
+// NewFallbackClient returns a new FallbackClient. l1ChainId and l1Block are used to check
+// whether the newly switched rpc is legal.
+func NewFallbackClient(ctx context.Context, rpc client.RPC, urlList []string, log log.Logger, l1ChainId *big.Int, l1Block eth.BlockID, rpcInitFunc func(url string) (client.RPC, error)) client.RPC {
+	fallbackClient := &FallbackClient{
+		ctx:          ctx,
+		firstRpc:     rpc,
+		urlList:      urlList,
+		log:          log,
+		rpcInitFunc:  rpcInitFunc,
+		currentRpc:   rpc,
+		currentIndex: 0,
+		l1ChainId:    l1ChainId,
+		l1Block:      l1Block,
+	}
+	go func() {
+		ticker := time.NewTicker(1 * time.Minute)
+		for {
+			select {
+			case <-ticker.C:
+				log.Debug("FallbackClient clear lastMinuteFail 0")
+				fallbackClient.lastMinuteFail.Store(0)
+			}
+		}
+	}()
+	return fallbackClient
+}
+
+func (l *FallbackClient) SubscribeNewHead(ctx context.Context, ch chan<- *types.Header) (ethereum.Subscription, error) {
+	return l.EthSubscribe(ctx, ch, "newHeads")
+}
+
+func (l *FallbackClient) Close() {
+	l.currentRpc.Close()
+	if l.currentRpc != l.firstRpc {
+		l.firstRpc.Close()
+	}
+}
+
+func (l *FallbackClient) CallContext(ctx context.Context, result any, method string, args ...any) error {
+	err := l.currentRpc.CallContext(ctx, result, method, args...)
+	if err != nil {
+		l.handleErr(err)
+	}
+	return err
+}
+
+func (l *FallbackClient) handleErr(err error) {
+	if err == rpc.ErrNoResult {
```

Comment:
Will there be other errors on the client side?

Reply:
Other errors will not be ignored, only the one that cannot find the result needs to be ignored.
Other possible values are as follows:
```golang
var (
	ErrBadResult                 = errors.New("bad result in JSON-RPC response")
	ErrClientQuit                = errors.New("client is closed")
	ErrNoResult                  = errors.New("no result in JSON-RPC response")
	ErrSubscriptionQueueOverflow = errors.New("subscription queue overflow")
	errClientReconnected         = errors.New("client reconnected")
	errDead                      = errors.New("connection lost")
)
```
Not sure if there are other values, due to the flexibility of golang, there may be other errors, such as `execution reverted`

Reply:
Added the following logic to rule out the case where the request is successful but the service returns an Error:
```golang
        if _, ok := err.(rpc.Error); ok {
		return
	}
```